ledfx/devices/hue.py

```diff
@@ -0,0 +1,158 @@
+            ): str,
+            vol.Optional("udp_port", description="port", default=2100): int,
+            vol.Required(
+                "user_name",
+                description="User name",
+            ): str,
+            vol.Required(
+                "client_key",
+                description="Client key",
+            ): str,
+            vol.Required(
+                "group_name",
+                description="Entertainment zone group name",
+            ): str,
+        }
+    )
+
+    status: Dict[int, Tuple[int, int, int]]
+    _sock: Optional[socket.socket] = None
+
+    def __init__(self, ledfx, config):
+        super().__init__(ledfx, config)
+
+        self._dtls_client_context = tls.ClientContext(
+            tls.DTLSConfiguration(
+                pre_shared_key=(
+                    self._config["user_name"],
+                    bytes.fromhex(self._config["client_key"]),
+                ),
+                ciphers=["TLS-PSK-WITH-AES-128-GCM-SHA256"],
+            )
+        )
+        self.status = {}
+
+    def _hue_get(self, api_endpoint, data=None):
+        return requests.get(
+            f"http://{self._config['ip_address']}/api/{self._config['user_name']}/{api_endpoint}",
+            json=data,
+        ).json()
+
+    def _hue_put(self, api_endpoint, data):
+        return requests.put(
+            f"http://{self._config['ip_address']}/api/{self._config['user_name']}/{api_endpoint}",
+            json=data,
+        ).json()
+
+    def _entertainment_groups(self):
+        all_groups = self._hue_get("groups")
+        return {
+            id: all_groups[id]
+            for id in all_groups
+            if all_groups[id]["type"] == "Entertainment"
+        }
+
+    def activate(self):
+        request_data = {"stream": {"active": True}}
+        response = self._hue_put(
+            f"groups/{self._config['group_id']}", request_data
+        )[0]
+        if "success" in response:
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            self._sock = self._dtls_client_context.wrap_socket(sock, None)
+            self._sock.connect(
+                (self._config["ip_address"], self._config["udp_port"])
+            )
+        else:
+            raise Exception("Unable to activate UDP stream mode")
+
+        # I'm a hack, but I seem to be necessary and I work
+        for _ in range(10):
+            try:
+                time.sleep(0.2)
+                self._sock.do_handshake()
+            except exceptions.TLSError as e:
+                ...
+
+        super().activate()
+
+    def deactivate(self):
+        if self._sock is not None:
+            self._sock.close()
+            self._sock = None
+        request_data = {"stream": {"active": False}}
+        self._hue_put(f"groups/{self._config['group_id']}", request_data)
+
+        super().deactivate()
+
+    def flush(self, data):
+        pixels = [[int(r), int(g), int(b)] for r, g, b in data]
```

Comment:
Since we're calling this 30 times a second for each bulb, we should benchmark if using floor is faster than casting to int - I'm suspicious it might be.

```py
pixels = [[r // 1, g // 1, b // 1] for r, g, b in data]
```

Reply:
This doesn't work.  I need an integer value, but because r, g, b are floats, the floor division returns a float.  That being said, I'm not seeing performance issues.